intel-sgx/ias/src/verifier.rs

```diff
@@ -0,0 +1,751 @@
+    }
+}
+
+/// Check that `cert` contains the `AttestationEmbeddedIasReport` X.509 extension and that the
+/// report in that extension:
+///
+///  * has a valid signature from a trusted report signer
+///  * reports a valid attestation
+///  * had REPORTDATA containing the hash of `cert`'s SubjectPublicKeyInfo
+///
+/// If the checks pass, returns `Ok(EnclavePeer { ... })` with the enclave identity from the
+/// validated report. If the checks do not pass, returns an error.
+///
+/// CAUTION: This routine does not verify the certificate signature nor the standard X.509
+/// attributes. The caller is responsible for that.
+pub fn verify_epid_cert_embedded_attestation<C: Crypto, P: Platform>(ca_certificates: &[&[u8]], cert: &GenericCertificate, platform_verifier: &P) -> Result<EnclavePeer, Error> {
+    let extn = cert.tbscert.get_extension(&oid::attestationEmbeddedIasReport)
+        .ok_or_else(|| Error::enclave_certificate(ErrorKind::MissingIasReport, None::<Error>))?;
+
+    // The extension is stored as a DER octetstring
+    let quote = AttestationEmbeddedIasReport::from_der(&extn.value)
+        .map_err(|e| Error::enclave_certificate(ErrorKind::ReportAsn1Parse, Some(e)))?
+        .quote::<C, _>(ca_certificates, platform_verifier)?;
+
+    // The SGX report data must contain the SPKI hash.
+    let mut expected_reportdata = [0; crypto::SHA256_DIGEST_LEN];
+    C::sha256(cert.tbscert.spki.as_ref(), &mut expected_reportdata)
+        .map_err(|e| Error::enclave_certificate(ErrorKind::SpkiHashFailure, Some(e)))?;
+    let mut expected_reportdata = Vec::from(expected_reportdata);
+    expected_reportdata.resize(::std::mem::size_of_val(&quote.reportdata), 0);
+
+    if expected_reportdata != &quote.reportdata[..] {
+        return Err(Error::enclave_certificate(ErrorKind::ReportBadPayload, None::<Error>));
+    }
+
+    Ok(quote.into())
+}
+
+pub fn is_supported(ias_version: IasVersion) -> bool {
+    let supported = SUPPORTED_IAS_VERSIONS.to_vec();
+    supported.contains(&ias_version)
+}
+
+pub trait Platform {
+    fn verify(&self, for_self: bool, nonce: &Option<String>, isv_enclave_quote_status: QuoteStatus, advisories: &Vec<IasAdvisoryId>) -> Result<(), Error>;
+}
+
+pub struct AutoDetect;
+
+impl AutoDetect {
+    pub fn new() -> AutoDetect {
+        AutoDetect {}
+    }
+}
+
+impl Platform for AutoDetect {
+    fn verify(&self, for_self: bool, _nonce: &Option<String>, isv_enclave_quote_status: QuoteStatus, advisories: &Vec<IasAdvisoryId>) -> Result<(), Error> {
+        // The report must be for a valid quote.
+        match (for_self, isv_enclave_quote_status) {
+            (false, QuoteStatus::SwHardeningNeeded) => {
+                /* We don't know how the enclave peer has been compiled. Enclaves may verify their own
+                 * attestations, so we can ignore `QuoteStatus::SwHardeningNeeded` here and rely on the
+                 * logic of enclave versions. It is of paramount importance that we only ignore
```

Comment:
We don't have much of a choice but to accept `SwHardeningNeeded` here. If that is undesirable, enclaves could self attest before doing any meaningful work; As long as they can't verify that they themselves are able to provide an acceptable attestation, they should refuse to operate. This self-attestation will execute the next arm in the match statement (`for_self = true`). I've tried to clarify this better in the comment.

Reply:
Thanks for the explanation and the additional comments in the code. I think I understand the reasoning now.

I'm still a bit worried about this approach, though, mainly because we're ignoring the `SwHardeningNeeded` status by default, without having the code that is requesting the report validation *explicitly* indicate that it is okay to do so. It feels like insecure default behavior to me. There is also no rust-doc on the the `AutoDetect` struct to warn about this.

I would say the minimum we should do is to document this behavior, to warn users about the security implications. We could also go a step further and require the code that is requesting the report validation to explicitly indicate how the `SwHardeningNeeded` status should be handled when verifying other (non-self) enclaves. For instance, we could refactor the `AutoDetect` struct into something more configurable, like:
```
// TODO: add proper documentation to this struct and to each of its member fields
pub struct DefaultPlatformVerifier {
    pub other_enclave_sw_hardening_needed_validation_strategy: IasAdvisoryStrategy,
   // other configuration values could be added here in the future, if needed
}

pub enum IasAdvisoryStrategy {
    IgnoreAll,
    Ignore(Vec<IasAdvisoryId>), // an empty vec would result in ignoring no advisories
}
```
Another alternative would be to remove the `AutoDetect` struct from the crate and instead keep it as something we should use internally only (with caution).

Reply:
Yes this is not ideal. There's some work being done internally on an alternative approach. I'll remove this `AutoDetect` implementation from the crate for now.